File: 01 CLASE ENTERO/js/entero.js
class Entero {
    //Atributos
    Num;

    //Constructor = inicializa el objeto
    constructor(Numero){
        this.Num = Numero;  
    }

    //CLASS- Cargar el valor de Num
    setNum(){
        const input = document.getElementById('numeroInput')
        this.Num = parseInt(input.value, 10)
    }


    //CLASS- Obtener el valor del numero 
    getNum() {
        return this.Num;    
    }

    //CLASS- Mostra el vaalor del numero
    showNum() {
        const resultadoDiv = document.getElementById('resultado')
        resultadoDiv.textContent = this.getNum();
    }

    showResultado(respuesta) {
        const resultadoDiv = document.getElementById('resultado')
        resultadoDiv.textContent = respuesta;
    }

    incrementarNum(){
        this.Num = this.Num + 1;
    }

    decrementarNum(){
        this.Num = this.Num - 1;
    }
 
    // Estos dos son lo mismo solo que resumido
    // esParImpar(){
    //     if (this.Num % 2 !== 0){
    //         return true;
    //     }else{
    //         return false
    //     }
    // }
    esParImpar(){
        return (this.Num % 2 == 0)
    }

    esPosiNega(){
        return (this.Num >= 0)
    }

    factorial(){
        if (this.Num < 0) return "No hay factorial de numero negativos";
        let factorial = 1, i = this.Num;
        while (i > 1) factorial *= i--;
        return factorial;
    }

    esPerfecto(){
        let suma = 0;  // Inicializamos una variable para almacenar la suma de divisores

        // Recorremos desde 1 hasta Num-1 para encontrar los divisores
        for (let i = 1; i < this.Num; i++) {
            if (this.Num % i === 0) {  // Si i es divisor de Num
                suma += i;  // Añadimos i a la suma
            }
        } 
        // Comprobamos si la suma de los divisores es igual al número
        return suma === this.Num;
    }

    esPrimo(){
        if (this.Num <= 1) return false;
        for (let i = 2; i <= Math.sqrt(this.Num); i++) {
            if (this.Num % i === 0) return false;
        }
        return true;
    }

    Fibonucci(){
        if (this.Num <= 0) return "No hay sucesión de Fibonacci para números menores o iguales a 0";
        
        let fibo = [0, 1];
        while (fibo[fibo.length - 1] + fibo[fibo.length - 2] <= this.Num) {
            fibo.push(fibo[fibo.length - 1] + fibo[fibo.length - 2]);
        }
        
        return fibo;
    }
}

// Las funciones = button HTML
var ClaseEntero = new Entero(0); //Se inicializo en 0

//crear el objeto 
function cargarNum() {
    ClaseEntero.setNum();
}

//mostrar el valor de num en html
function mostrarNum(){
    ClaseEntero.showNum();
}

function incrementarValor(){
    ClaseEntero.incrementarNum();
    ClaseEntero.showNum();
}

function decrementarValor(){
    ClaseEntero.decrementarNum();
    ClaseEntero.showNum();
}

function esParImparNum(){
    var respuesta = ClaseEntero.esParImpar();
    var resp = respuesta ? "Es Num Par":"Es Num Impar";
    ClaseEntero.showResultado(resp)
}

function esPosiNegaNum(){
    var respuesta = ClaseEntero.esPosiNega();
    var resp = respuesta ? "Es Num Positivo":"Es Num Negativo";
    ClaseEntero.showResultado(resp)
}

function factorialNum(){
    var respuesta = ClaseEntero.factorial();
    ClaseEntero.showResultado("Factorial:" + respuesta)
}

function esPerfectoNum(){
    var respuesta = ClaseEntero.esPerfecto();
    var resp = respuesta ? "Es Num perfecto":"No es Num perfecto"
    ClaseEntero.showResultado(resp)
}

function esPrimoNum(){
    var respuesta = ClaseEntero.esPrimo();
    var resp = respuesta ? "Es Num Primo" : "No es Num Primo";
    ClaseEntero.showResultado(resp);
}

function FibonucciNum(){
    var respuesta = ClaseEntero.Fibonucci();
    ClaseEntero.showResultado("Fibonacci: " + respuesta.join(", "));
}

File: 01 CLASE ENTERO/py/entero.py
class Entero:
    def __init__(self,Numero):
        self.Num = Numero

    def setNum(self):
        input_value = input("Enter a number: ")
        self.Num = int(input_value)

    def getNum(self):
        return self.Num
    
    def showNum(self):
        print(self.getNum())

    def showResultado(self, respuesta):
        print(respuesta)

    def incremnetarNum(self):
        self.Num += 1

    def decremnetarNum(self):
        self.Num -= 1 
    
    def esParImpar(self):
        return (self.Num % 2 == 0)
    
    def esPerfecto(self):
        suma_divisores = sum([i for i in range(1, self.Num) if self.Num % i == 0])
        return suma_divisores == self.Num

    def esPrimo(self):
        if self.Num <= 1:
            return False
        for i in range(2, int(self.Num ** 0.5) + 1):
            if self.Num % i == 0:
                return False
        return True

    def fibonacci(self):
        fibo = [0, 1]
        while fibo[-1] < self.Num:
            fibo.append(fibo[-1] + fibo[-2])
        if fibo[-1] > self.Num:
            fibo.pop()  # Eliminar el último número si supera a self.Num
        return fibo
    
    def menu(self):
        while True:
            print("\nMenu:")
            print("1. Establecer un nuevo numero")
            print("2. Mostrar el numero actual")
            print("3. Incrementar el numero")
            print("4. Decrementar el numero")
            print("5. Verificar si es par o impar")
            print("6. Verificar si es un número perfecto")
            print("7. Verificar si es un número primo")
            print("8. Mostrar la sucesión de Fibonacci")
            print("9. Salir")

            opcion = input("Ingrese una opcion: ")

            if opcion == '1':
                self.setNum()
            elif opcion == '2':
                self.showNum()
            elif opcion == '3':
                self.incremnetarNum()
            elif opcion == '4':
                self.decremnetarNum()
            elif opcion == '5':
                if self.esParImpar():
                    print("El numero es par")
                else:
                    print("El numero es impar")
            elif opcion == '6':
                if self.esPerfecto():
                    print("El numero es perfecto")
                else:
                    print("El numero no es perfecto")
            elif opcion == '7':
                if self.esPrimo():
                    print("El numero es primo")
                else:
                    print("El numero no es primo")
            elif opcion == '8':
                print(f"La sucesión de Fibonacci hasta {self.Num} es: {self.fibonacci()}")
            elif opcion == '9':
                print ("Hasta luego!")
                break
            else:
                print("Opcion invalida.Intente nuevamente")
    

numero = Entero(10)

numero.menu()

